[-] Smart Field Search - Ease the search on smart fields and deprecate old ways to do it

VincentMolinie opened the pull request:
## Pull Request checklist:

- [ ] Write an explicit title for the Pull Request
- [ ] Write changes made in the CHANGELOG.md
- [ ] Create automatic tests
- [ ] Test manually the implemented changes
- [ ] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code


## Before
We had to do this kind of thing
```javascript
collection('customers', {
  fields: [{
    field: 'fullname',
    type: 'String',
    get: (customer) => {
      return customer.firstname + ' ' + customer.lastname;
    },
    search: function (query, search) {
      let s = models.sequelize;
      let split = search.split(' ');

      var searchCondition = {
        [Op.and]: [
          { firstname: { [Op.like]: `%${split[0]}%` } },
          { lastname: { [Op.like]: `%${split[1]}%` } },
        ]
      };

      query.where[Op.and][0][Op.or].push(searchCondition);

      return query;
    }
  }]
});
```

## After
```javascript
collection('customers', {
  fields: [{
    field: 'fullname',
    type: 'String',
    get: (customer) => {
      return customer.firstname + ' ' + customer.lastname;
    },
    search: function (query, search) {
      let s = models.sequelize;
      let split = search.split(' ');

      return {
        [Op.and]: [
          { firstname: { [Op.like]: `%${split[0]}%` } },
          { lastname: { [Op.like]: `%${split[1]}%` } },
        ]
      };
    }
  }]
});
```

SteveBunlon commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 
 ### Changed
 - Technical - Upgrade to babel 7 stable.
+- Smart Field Search - Ease the search on smart fields and deprecate old ways to do it.
```
[Thinking] IMO this is not explicit enough. Something like : 
`Ease the search on smart fields by requiring a condition instead of a query, and deprecate old way to do it.` Give a little bit more details

SteveBunlon commented on src/services/has-many-getter.js:
```diff
@@ -22,10 +22,18 @@ function HasManyGetter(model, association, opts, params) {
+      where = await searchBuilder.perform(params.associationName);
+    }
+    return where;
+  }
```
[IMPROVEMENT] I don't like that, the `where` value should initialized as the other, no need to create a function.

VincentMolinie replied:
It's async. So I cannot initialized it like the other 😢

SteveBunlon commented on src/services/has-many-getter.js:
```diff
@@ -38,20 +46,20 @@ function HasManyGetter(model, association, opts, params) {
         scope: false,
         required: false,
-        where,
+        where: getWhere(),
```
[MISC] This function returns a promise, the `where` operator expects an object not a promise. This won't work.

SteveBunlon commented on src/services/has-many-getter.js:
```diff
@@ -38,20 +46,20 @@ function HasManyGetter(model, association, opts, params) {
   }
 
-  function getCount() {
+  async function getCount() {
```
[SCARED] You are returning a promise of promise (wait for the count value and return it directly instead of returning a promise of the value) and you don't wait for the 'getWhere()' promise value here again. The `where` operator expects an object.

VincentMolinie replied:
A promise of a promise is a promise 🤔

SteveBunlon commented on src/services/search-builder.js:
```diff
@@ -69,20 +73,54 @@ function SearchBuilder(model, opts, params, fieldNamesRequested) {
+      if (field.isVirtual) {
+        if (field.search) {
+          try {
+            const queryForRetrocompatibility = {
```
```suggestion
            const queryForBackwardCompatibility = {
```

SteveBunlon commented on src/services/search-builder.js:
```diff
@@ -69,20 +73,54 @@ function SearchBuilder(model, opts, params, fieldNamesRequested) {
+              },
+            };
+            condition = await Promise.resolve(
+              field.search(queryForRetrocompatibility, params.search),
```
```suggestion
              field.search(queryForBackwardCompatibility, params.search),
```

SteveBunlon commented on src/services/search-builder.js:
```diff
@@ -57,8 +60,9 @@ function SearchBuilder(model, opts, params, fieldNamesRequested) {
   this.getFieldsSearched = () => fieldsSearched;
 
   this.hasExtendedSearchConditions = () => hasExtendedConditions;
+  this.hasSearchOnSmartField = () => hasSmartFieldSearch;
```
[MISC] Why do we export a boolean as a function ?

arnaudbesnier commented:
> ## Pull Request checklist:
> * [x]  Write an explicit title for the Pull Request
> * [x]  Write changes made in the CHANGELOG.md
> * [ ]  Create automatic tests
> * [ ]  Test manually the implemented changes
> * [ ]  Review my own code (indentation, syntax, style, simplicity, readability)
> * [ ]  Wonder if you can improve the existing code
> 
> ## Before
> We had to do this kind of thing
> 
> ```js
> collection('customers', {
>   fields: [{
>     field: 'fullname',
>     type: 'String',
>     get: (customer) => {
>       return customer.firstname + ' ' + customer.lastname;
>     },
>     search: function (query, search) {
>       let s = models.sequelize;
>       let split = search.split(' ');
> 
>       var searchCondition = {
>         [Op.and]: [
>           { firstname: { [Op.like]: `%${split[0]}%` } },
>           { lastname: { [Op.like]: `%${split[1]}%` } },
>         ]
>       };
> 
>       query.where[Op.and][0][Op.or].push(searchCondition);
> 
>       return query;
>     }
>   }]
> });
> ```
> 
> ## After
> ```js
> collection('customers', {
>   fields: [{
>     field: 'fullname',
>     type: 'String',
>     get: (customer) => {
>       return customer.firstname + ' ' + customer.lastname;
>     },
>     search: function (query, search) {
>       let s = models.sequelize;
>       let split = search.split(' ');
> 
>       return {
>         [Op.and]: [
>           { firstname: { [Op.like]: `%${split[0]}%` } },
>           { lastname: { [Op.like]: `%${split[1]}%` } },
>         ]
>       };
>     }
>   }]
> });
> ```

So you can't use `include` to join if necessary?

VincentMolinie commented:
> > ## Pull Request checklist:
> > 
> > * [x]   Write an explicit title for the Pull Request
> > * [x]   Write changes made in the CHANGELOG.md
> > * [ ]   Create automatic tests
> > * [ ]   Test manually the implemented changes
> > * [ ]   Review my own code (indentation, syntax, style, simplicity, readability)
> > * [ ]   Wonder if you can improve the existing code
> > 
> > ## Before
> > We had to do this kind of thing
> > ```js
> > collection('customers', {
> >   fields: [{
> >     field: 'fullname',
> >     type: 'String',
> >     get: (customer) => {
> >       return customer.firstname + ' ' + customer.lastname;
> >     },
> >     search: function (query, search) {
> >       let s = models.sequelize;
> >       let split = search.split(' ');
> > 
> >       var searchCondition = {
> >         [Op.and]: [
> >           { firstname: { [Op.like]: `%${split[0]}%` } },
> >           { lastname: { [Op.like]: `%${split[1]}%` } },
> >         ]
> >       };
> > 
> >       query.where[Op.and][0][Op.or].push(searchCondition);
> > 
> >       return query;
> >     }
> >   }]
> > });
> > ```
> > 
> > 
> > ## After
> > ```js
> > collection('customers', {
> >   fields: [{
> >     field: 'fullname',
> >     type: 'String',
> >     get: (customer) => {
> >       return customer.firstname + ' ' + customer.lastname;
> >     },
> >     search: function (query, search) {
> >       let s = models.sequelize;
> >       let split = search.split(' ');
> > 
> >       return {
> >         [Op.and]: [
> >           { firstname: { [Op.like]: `%${split[0]}%` } },
> >           { lastname: { [Op.like]: `%${split[1]}%` } },
> >         ]
> >       };
> >     }
> >   }]
> > });
> > ```
> 
> So you can't use `include` to join if necessary?

You are absolutely right. I thought about it, but most of the time includes does not work as expected (For example try with a hasMany...).

I think we need to rework a little bit the smart field to add a `requirement` property for example
That field could look like `requirement: ['users.*', 'companies.{id,name}']`...

Tell me what you think about it @arnaudbesnier

arnaudbesnier commented:
> > > ## Pull Request checklist:
> > > 
> > > * [x]    Write an explicit title for the Pull Request
> > > * [x]    Write changes made in the CHANGELOG.md
> > > * [ ]    Create automatic tests
> > > * [ ]    Test manually the implemented changes
> > > * [ ]    Review my own code (indentation, syntax, style, simplicity, readability)
> > > * [ ]    Wonder if you can improve the existing code
> > > 
> > > ## Before
> > > We had to do this kind of thing
> > > ```js
> > > collection('customers', {
> > >   fields: [{
> > >     field: 'fullname',
> > >     type: 'String',
> > >     get: (customer) => {
> > >       return customer.firstname + ' ' + customer.lastname;
> > >     },
> > >     search: function (query, search) {
> > >       let s = models.sequelize;
> > >       let split = search.split(' ');
> > > 
> > >       var searchCondition = {
> > >         [Op.and]: [
> > >           { firstname: { [Op.like]: `%${split[0]}%` } },
> > >           { lastname: { [Op.like]: `%${split[1]}%` } },
> > >         ]
> > >       };
> > > 
> > >       query.where[Op.and][0][Op.or].push(searchCondition);
> > > 
> > >       return query;
> > >     }
> > >   }]
> > > });
> > > ```
> > > 
> > > 
> > > ## After
> > > ```js
> > > collection('customers', {
> > >   fields: [{
> > >     field: 'fullname',
> > >     type: 'String',
> > >     get: (customer) => {
> > >       return customer.firstname + ' ' + customer.lastname;
> > >     },
> > >     search: function (query, search) {
> > >       let s = models.sequelize;
> > >       let split = search.split(' ');
> > > 
> > >       return {
> > >         [Op.and]: [
> > >           { firstname: { [Op.like]: `%${split[0]}%` } },
> > >           { lastname: { [Op.like]: `%${split[1]}%` } },
> > >         ]
> > >       };
> > >     }
> > >   }]
> > > });
> > > ```
> > 
> > 
> > So you can't use `include` to join if necessary?
> 
> You are absolutely right. I thought about it, but most of the time includes does not work as expected (For example try with a hasMany...).
> 
> I think we need to rework a little bit the smart field to add a `requirement` property for example
> That field could look like `requirement: ['users.*', 'companies.{id,name}']`...
> 
> Tell me what you think about it @arnaudbesnier

I am not ok to remove flexibility here.

The day developers will update to the new (major?) version containing this code, they will potentially lose their Smart Field search feature (if they used `include`). There is no alternatives with the current code.

I am not fan of the `requirement` attribute idea as presented, because:
- it is defined in the field but only impacts the `search` behaviour.
- from my point of view, all this search configuration/implementation should be located in the `search` function.

VincentMolinie commented:
Actually after thinking about it the include is useless as of right now the call is async so you can  do your search on the other model if necessary and then do a id in [myListOfId] for example

arnaudbesnier commented:
Ok, I think this something could also have with the current implementation.

What is important is what is removed.
With the new implementation, you won't have the ability to do it in a single SQL query anymore.

VincentMolinie commented:
> Ok, I think this something could also have with the current implementation.
> 
> What is important is what is removed.
> With the new implementation, you won't have the ability to do it in a single SQL query anymore.

yeah but like I said it won't work most of the time, because if you try to do it on `many` relationships it won't work. And belongsTo might work but not like you might expect...
